---
title: "Angeletti Data Analysis Report"
author: "Jonathan Robinson"
date: "`r Sys.Date()`"
output:
  html_document:
    fig_caption: yes
    toc: no
  pdf_document:
    toc: yes
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F, results = 'hide', fig.width=10)
```

```{r}
# specify relevant directory information
pdir <- '/Users/jonrob/Documents/NBIS/LTS_projects/2020_DAngeletti/'  # parent directiory
adir <- paste0(pdir,'analysis/')  # analysis subdirectory
ddir <- paste0(pdir,'data/')  # data subdirectory
```

```{r}
# load libraries
library(knitr)
```

# Report {.tabset}

## Data summary

A summary of the data samples included in this analysis report.

```{r, results='asis'}
meta <- read.csv2(paste0(ddir,'metadata.csv'))
kable(meta, caption='scRNAseq metadata')
```


## Quality control {.tabset}

### Gene QC

##### Gene Biotype proportions

Here we investigated the proportion of different gene **biotypes** encountered in the dataset.

```{r,results='asis', echo=FALSE, fig.width=10}
files <- '01_qc/gene_biotype_proportions.png'
for (i in files){ cat("![](",adir,i,")", sep='') }
```
**Figure.** Proportion of gene biotypes in the dataset.

##### Gene Family proportion

Here we investigated the proportion of different gene **families** encountered in the dataset.

```{r,results='asis', echo=FALSE, fig.width=10}
files <- '01_qc/Gene_familty proportions.png'
for (i in files){ cat("![](",adir,i,")", sep='') }
```
**Figure.** Proportion of gene families in the dataset.


### QC plots BEFORE filtering
#### Samples grouped by INFECTION STATUS
```{r,results='asis'}
cat('![](', adir, '01_qc/QC_infection_ALL.png', ')', sep='')
```

#### Samples grouped by ORGAN
```{r,results='asis'}
cat('![](', adir, '01_qc/QC_organ_ALL.png', ')', sep='')
```

#### Samples grouped by DAY POST INFECTION
```{r,results='asis'}
cat('![](', adir, '01_qc/QC_day_post_infection_ALL.png', ')', sep='')
```


### QC plots AFTER filtering
#### Samples grouped by INFECTION STATUS
```{r,results='asis'}
cat('![](', adir, '01_qc/QC_infection_FILTERED.png', ')', sep='')
```

#### Samples grouped by ORGAN
```{r,results='asis'}
cat('![](', adir, '01_qc/QC_organ_FILTERED.png', ')', sep='')
```

#### Samples grouped by DAY POST INFECTION
```{r,results='asis'}
cat('![](', adir, '01_qc/QC_day_post_infection_FILTERED.png', ')', sep='')
```


## Dimensionality reduction {.tabset}

```{r}
cdir <- paste0(adir, '06_cluster/')
```

### Metadata (factors)
```{r, results='asis'}
metadata_fields <- c('organ','infection','day_post_infection')
for (mf in metadata_fields){
  cat('#### UMAP: colored by', unlist(strsplit(mf,'_')))
  cat('![](', cdir, 'umap_plots/umap_metadata_', mf, '.png', '){width=80%}\n', sep='')
}
```

### Metadata (continuous)
```{r, results='asis'}
metadata_fields <- c('nCount_RNA','nFeature_RNA','S_Score')
for (mf in metadata_fields){
  # cat('#### UMAP: colored by', unlist(strsplit(mf,'_')))
  cat('![](', cdir, 'umap_plots/umap_metadata_', mf, '.png', '){width=80%}\n', sep='')
}
```


## Clustering

```{r}
clust <- 'HC_11'
```

```{r, results='asis'}
cat('#### UMAP with mapped hierarchical cluster groups (', clust, ')', sep='')
cat('![](', cdir, 'clustering/clustering_', clust, '_umap.png', '){width=80%}', sep='')
```


## Differential expression {.tabset}
```{r}
edir <- paste0(adir,'07_diff_expr/')
```

### Violin plots
#### Showing the top 3 DE genes from each cluster
```{r, results='asis'}
cat('![](', edir, 'violinPlot_genes_per_cluster_top3.png', ')', sep='')
```

### Heatmap
#### Showing the top 3 DE genes from each cluster
```{r, results='asis'}
cat('![](', edir, 'heatmap_genes_per_cluster_top3.png', ')', sep='')
```


## VDJ Analysis {.tabset}

```{r}
# specify levels of analysis/output
analysis_types <- c('unpaired', 'paired')
chains <- c('IGH', 'IGK', 'IGL', 'IGH-IGK', 'IGH-IGL')
chainparts <- c('cdr3','cdr3_nt','v_gene','j_gene','merged')
```


```{r, results='asis'}
for (atype in analysis_types) {
  
  cat('###', atype, '{.tabset}\n')
  vdir <- paste0(adir, '08_VDJ_analysis_', atype, '/')
  
  for (chain in chains) {
    
    cat('####', chain, '{.tabset}\n')
    
    # generate directory names
    if (chain == 'IGH-IGK') {
      chaindir <- 'IGH|IGK'
      cparts <- 'IGhk_clonotype'
    } else if (chain == 'IGH-IGL') {
      chaindir <- 'IGH|IGL'
      cparts <- 'IGhl_clonotype'
    } else {
      chaindir <- chain
      cparts <- chainparts
    }
    
    # show plots for each chain and chain piece
    for (cpart in cparts) {
      cat('#####', cpart, '\n')
      chain_cpart <- paste0(chaindir, '_', cpart)
      chain_subdir <- paste0(vdir, chain_cpart, '/')
      
      cat('![](', chain_subdir, chain_cpart, '_diversity_plots.png', ')\n', sep='')
      if (cpart %in% c('v_gene', 'j_gene')) {
        
      }
      cat('![](', chain_subdir, chain_cpart, '_umap_scale.png', ')\n\n', sep='')
    }
  }
}
```




