---
title: "scRNAseq Data Analysis Report"
author: "Jonathan Robinson"
date: "`r Sys.Date()`"
output:
  html_document:
    fig_caption: yes
    toc: no
  pdf_document:
    toc: yes
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F, results = 'hide', fig.width=10)
```

```{r}
# specify relevant directory information
pdir <- '/Users/jonrob/Documents/NBIS/LTS_projects/d_angeletti_1910/scMouseBcellFlu/'  # parent directiory
adir <- paste0(pdir,'analysis/')  # analysis subdirectory
ddir <- paste0(pdir,'data/')  # data subdirectory
```

```{r}
# load libraries
library(knitr)
```

# {.tabset}

## Data summary

A summary of the data samples included in this analysis report.

```{r, results='asis'}
meta <- read.csv(paste0(ddir,'metadata.csv'))
kable(meta, caption='scRNAseq metadata')
```


## Quality control {.tabset}

### Gene QC

##### Gene Biotype proportions

Here we investigated the proportion of different gene **biotypes** encountered in the dataset.

```{r,results='asis', echo=FALSE, fig.width=10}
files <- '01_qc/gene_biotype_proportions.png'
for (i in files){ cat("![](",adir,i,")", sep='') }
```
**Figure.** Proportion of gene biotypes in the dataset.

##### Gene Family proportion

Here we investigated the proportion of different gene **families** encountered in the dataset.

```{r,results='asis', echo=FALSE, fig.width=10}
files <- '01_qc/Gene_familty proportions.png'
for (i in files){ cat("![](",adir,i,")", sep='') }
```
**Figure.** Proportion of gene families in the dataset.


### QC plots BEFORE filtering
#### Samples grouped by INFECTION STATUS
```{r,results='asis'}
cat('![](', adir, '01_qc/QC_infection_ALL.png', ')', sep='')
```

#### Samples grouped by ORGAN
```{r,results='asis'}
cat('![](', adir, '01_qc/QC_organ_ALL.png', ')', sep='')
```


### QC plots AFTER filtering
#### Samples grouped by INFECTION STATUS
```{r,results='asis'}
cat('![](', adir, '01_qc/QC_infection_FILTERED.png', ')', sep='')
```

#### Samples grouped by ORGAN
```{r,results='asis'}
cat('![](', adir, '01_qc/QC_organ_FILTERED.png', ')', sep='')
```

## B-cell purification {.tabset}
```{r}
cdir <- paste0(adir, '02_cluster/')
```

### Clustering
```{r}
clust <- 'louvain_0.65'
```

```{r, results='asis'}
cat('#### UMAP with mapped cluster groups (', clust, ')', sep='')
cat('![](', cdir, 'clustering/clustering_', clust, '_umap.png', '){width=80%}', sep='')
```


### Differential expression {.tabset}
```{r}
edir <- paste0(adir,'03_diff_expr/')
```

#### Violin plots
##### Showing the top 3 DE genes from each cluster
```{r, results='asis'}
cat('![](', edir, 'violinPlot_genes_per_cluster_top3.png', ')', sep='')
```

#### Heatmap
##### Showing the top 3 DE genes from each cluster
```{r, results='asis'}
cat('![](', edir, 'heatmap_genes_per_cluster_top3.png', ')', sep='')
```


### Cell type prediction
```{r}
pred_dir <- paste0(cdir, 'cell_type_prediction/main_cell_types/')
```

#### Predicted cell types (based on correlation)
```{r, results='asis'}
cat('![](', pred_dir, 'cell_cluster_pred_correlation_umap.png', '){width=70%}', sep='')
```

#### Predicted cell type composition of each cluster
```{r, results='asis'}
cat('![](', cdir, 'clustering/clustering_', clust, '_umap.png', '){width=40%}', sep='')
cat('![](', pred_dir, 'cell_cluster_pred_correlation_barplot.png', '){width=60%}', sep='')
```

<br/><br/>

#### Table of cell-type marker genes used for prediction
```{r, results='asis'}
gene_table <- read.csv(paste0(ddir,'gene_lists/main_cell_types.csv'))
kable(gene_table)
```


### Filtered cells
#### Predicted non-B-cells removed from the data
```{r, results='asis'}
cat('![](', adir, '04_remove_cells/umap_removed_cells.png', '){width=80%}', sep='')
```


## Cell/cluster features {.tabset}

```{r}
cdir <- paste0(adir, '06_cluster/')
```

### Metadata (factors)
```{r, results='asis'}
metadata_fields <- c('organ','day_post_infection','mouse_nr')
for (mf in metadata_fields){
  cat('#### UMAP: colored by', unlist(strsplit(mf,'_')))
  cat('![](', cdir, 'umap_plots/umap_metadata_', mf, '.png', '){width=80%}\n', sep='')
}
```

### Metadata (continuous)
#### UMAP colored by RNA counts and percent mitochondrial reads
```{r, results='asis'}
cat('![](', cdir, 'umap_plots/umap_metadata_nCount_RNA_perc_mito.png', ')', sep='')
```

<br/><br/>

#### UMAP colored by cell cycle scores
```{r, results='asis'}
cat('![](', cdir, 'umap_plots/umap_metadata_cellcycle.png', ')', sep='')
```

### Clustering
```{r}
clust <- 'HC_16'
```

```{r, results='asis'}
cat('##### UMAP with mapped cluster groups (', clust, ')\n', sep='')
cat('Note this is just one of the many available clustering methods/results. This grouping is shown because it was used for this differential expression analysis.\n\n')
cat('![](', cdir, 'clustering/clustering_', clust, '_umap.png', '){width=80%}', sep='')
```

### Differential expression {.tabset}
```{r}
edir <- paste0(adir,'07_diff_expr/')
```

#### Violin plots
##### Showing the top 3 DE genes from each cluster
```{r, results='asis'}
cat('![](', edir, 'violinPlot_genes_per_cluster_top3.png', ')', sep='')
```

#### Heatmap
##### Showing the top 3 DE genes from each cluster
```{r, results='asis'}
cat('![](', edir, 'heatmap_genes_per_cluster_top3.png', ')', sep='')
```


### Cell subtypes {.tabset}
```{r}
sdir <- paste0(cdir, '/cell_type_prediction/')
```

#### B-cell types
##### Correlation of cells with B-cell subtype marker genes
```{r, results='asis'}
cat('![](', sdir, 'bcell_types/UMAP_cell_type_correlation.png', ')', sep='')
```

<br/><br/>

##### Table of cell-type marker genes used for prediction
```{r, results='asis'}
gene_table <- read.csv(paste0(ddir,'gene_lists/bcell_types.csv'))
kable(gene_table)
```


#### Germinal center subtypes
##### Correlation of cells with B-cell GC subtype marker genes
```{r, results='asis'}
cat('![](', sdir, 'bcell_types_germsub_zonesub/UMAP_GC_cell_type_correlation.png', ')', sep='')
```

<br/><br/>

##### Table of cell-type marker genes used for prediction
```{r, results='asis'}
gene_table <- read.csv(paste0(ddir,'gene_lists/bcell_types_germsub_zonesub.csv'))
gene_table <- gene_table[, startsWith(colnames(gene_table), 'GC')]
kable(gene_table)
```






<!-- ## VDJ Analysis {.tabset} -->

<!-- ```{r} -->
<!-- # specify levels of analysis/output -->
<!-- analysis_types <- c('unpaired', 'paired') -->
<!-- chains <- c('IGH', 'IGK', 'IGL', 'IGH-IGK', 'IGH-IGL') -->
<!-- chainparts <- c('cdr3','cdr3_nt','v_gene','j_gene','merged') -->
<!-- ``` -->


<!-- ```{r, results='asis'} -->
<!-- for (atype in analysis_types) { -->

<!--   cat('###', atype, '{.tabset}\n') -->
<!--   vdir <- paste0(adir, '06_VDJ_analysis_', atype, '/') -->

<!--   for (chain in chains) { -->

<!--     cat('####', chain, '{.tabset}\n') -->

<!--     # generate directory names -->
<!--     if (chain == 'IGH-IGK') { -->
<!--       chaindir <- 'IGH|IGK' -->
<!--       cparts <- 'IGhk_clonotype' -->
<!--     } else if (chain == 'IGH-IGL') { -->
<!--       chaindir <- 'IGH|IGL' -->
<!--       cparts <- 'IGhl_clonotype' -->
<!--     } else { -->
<!--       chaindir <- chain -->
<!--       cparts <- chainparts -->
<!--     } -->

<!--     # show plots for each chain and chain piece -->
<!--     for (cpart in cparts) { -->
<!--       cat('#####', cpart, '\n') -->
<!--       chain_cpart <- paste0(chaindir, '_', cpart) -->
<!--       chain_subdir <- paste0(vdir, chain_cpart, '/') -->

<!--       cat('![](', chain_subdir, chain_cpart, '_diversity_plots.png', ')\n', sep='') -->
<!--       if (cpart %in% c('v_gene', 'j_gene')) { -->

<!--       } -->
<!--       cat('![](', chain_subdir, chain_cpart, '_umap_scale.png', ')\n\n', sep='') -->
<!--     } -->
<!--   } -->
<!-- } -->
<!-- ``` -->


